### PR TITLE
[FLINK-10230][table] Support 'SHOW CREATE VIEW' syntax to print the query of a view

### PR DESCRIPTION
**Pull request by RocMarshal**
## What is the purpose of the change

*Support 'SHOW CREATE VIEW' syntax to print the query of a view*

## Brief change log

*Support 'SHOW CREATE VIEW' syntax to print the query of a view*

## Verifying this change

This change added tests and can be verified as follows:
- org.apache.flink.table.planner.catalog.CatalogViewITCase

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / **docs** / JavaDocs / not documented)

### CODE REVIEW

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java [hunk 1]
```diff
@@ -1441,6 +1468,59 @@ private String buildShowCreateTableRow(
...
+        } else {
+            stringBuilder.append(
+                    String.format(
+                            "CREATE %sVIEW %s%s as%s%s",
+                            isTemporary ? "TEMPORARY " : "",
+                            viewIdentifier.asSerializableString(),
+                            String.format(
+                                    "(%s)",
+                                    schema.getColumns().stream()
+                                            .map(Column::getName)
```
Comment (Airblader) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java:1511]:
```suggestion
                                            .map(Column::getName)
                                            .map(EncodingUtils::escapeIdentifier)
```

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/catalog/CatalogViewITCase.scala [hunk 2]
```diff
@@ -18,24 +18,24 @@
...
 import org.junit.runner.RunWith
 import org.junit.runners.Parameterized
 import org.junit.{Before, Rule, Test}
 
 import java.util
-
 import scala.collection.JavaConversions._
 
-/** Test cases for view related DDLs. */
+/** Test cases for view related DDLs. TODO roc todo */
```
Comment (Airblader) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/catalog/CatalogViewITCase.scala:38]: TODO?

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlShowCreate.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+package org.apache.flink.sql.parser.dql;
+
+import org.apache.calcite.sql.SqlCall;
+import org.apache.calcite.sql.SqlIdentifier;
+import org.apache.calcite.sql.parser.SqlParserPos;
+
+/** Abstract class for {@link SqlShowCreateTable} and {@link SqlShowCreateView}. */
+public abstract class SqlShowCreate extends SqlCall {
+
+    protected SqlIdentifier sqlIdentifier;
```
Comment (Airblader) [inline flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlShowCreate.java:28]:
```suggestion
    protected final SqlIdentifier sqlIdentifier;
```

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java [hunk 4]
```diff
@@ -1176,6 +1178,31 @@ public TableResult executeInternal(Operation operation) {
                                         .getTableIdentifier()
                                         .asSerializableString()));
             }
+        } else if (operation instanceof ShowCreateViewOperation) {
+            ShowCreateViewOperation showCreateViewOperation = (ShowCreateViewOperation) operation;
+            Optional<CatalogManager.TableLookupResult> result =
+                    catalogManager.getTable(showCreateViewOperation.getViewIdentifier());
+            if (result.isPresent()) {
```
Comment (Airblader) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java:1185]: We can move this check into the Optional, which will save us from one level of indentation:

```
final CatalogManager.TableLookupResult result = catalogManager
    .getTable(showCreateViewOperation.getViewIdentifier())
    .orElseThrow(() -> new ValidationException(…));

return TableResultImpl.builder()…
```

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java [hunk 5]
```diff
@@ -1441,6 +1468,59 @@ private String buildShowCreateTableRow(
         return sb.toString();
     }
 
+    /** Show create view statement only for views. */
+    private String buildShowCreateViewRow(
```
Comment (Airblader) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java:1472]: This method is largely a duplicate of `buildShowCreateTableRow`. It'd be nice to refactor this a bit to extract some of the common logic, i.e. to format the columns.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java [hunk 6]
```diff
@@ -1441,6 +1468,59 @@ private String buildShowCreateTableRow(
...
+                                    schema.getColumns().stream()
+                                            .map(Column::getName)
+                                            .collect(Collectors.joining(", "))),
+                            System.lineSeparator(),
+                            ((CatalogView) view.getOrigin()).getOriginalQuery()));
+        }
+        // append comment
+        String comment = view.getComment();
+        if (StringUtils.isNotEmpty(comment)) {
+            stringBuilder.append(String.format(" COMMENT '%s'", comment));
```
Comment (Airblader) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java:1519]:
```suggestion
            stringBuilder.append(String.format(" COMMENT '%s'", EncodingUtils.escapeSingleQuotes(comment)));
```

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/catalog/CatalogViewITCase.scala [hunk 7]
```diff
@@ -307,6 +307,91 @@ class CatalogViewITCase(isStreamingMode: Boolean) extends AbstractTestBase {
     // now we only have permanent view T3
     assertEquals(permanentViewData.sorted, TestCollectionTableFactory.RESULT.sorted)
   }
+
+  @Test
+  def testShowCreateView(): Unit = {
+    val sourceData = List(toRow(1, "1000", 2))
+    TestCollectionTableFactory.initData(sourceData)
```
Comment (Airblader) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/catalog/CatalogViewITCase.scala:314]: We have new infrastructure for creating source/sink tables that don't require these big runtime implementation sources, see `TableFactoryHarness`. You could use that and create the table using `TableDescriptor` / `tableEnv.createTable`.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/catalog/CatalogViewITCase.scala [hunk 8]
```diff
@@ -307,6 +307,91 @@ class CatalogViewITCase(isStreamingMode: Boolean) extends AbstractTestBase {
...
+    val tView2DDL = "CREATE TEMPORARY VIEW t_v2(d, e, f) AS SELECT a, b, c FROM T1"
+    val view1DDL = "CREATE VIEW v1 AS SELECT a, b, c FROM T1"
+    val view2DDL = "CREATE VIEW v2(x, y, z) AS SELECT a, b, c FROM T1"
+    tableEnv.executeSql(tableDDL)
+    tableEnv.executeSql(tView1DDL)
+    tableEnv.executeSql(tView2DDL)
+    tableEnv.executeSql(view1DDL)
+    tableEnv.executeSql(view2DDL)
+    testShowCreateTemporaryView(tableEnv)
+    testShowCreateView(tableEnv)
```
Comment (Airblader) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/catalog/CatalogViewITCase.scala:333]: The split of methods here is really odd, because none of these methods are actually parameterized. I think this would be nicer to read if we organize the test like this:

```
// … setup code …

// Temporary views
// … code to set up first view …
// … assert first view …

// … code to set up second view …
// … assert second view …

// Catalog views
// … code to set up first view …
// … assert first view …

// … code to set up second view …
// … assert second view …
```

If anything, we could split it between temporary and catalog views (into two tests), that'd be a logical way to split this up a bit.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java [hunk 9]
```diff
@@ -1441,6 +1468,59 @@ private String buildShowCreateTableRow(
...
+    private String buildShowCreateViewRow(
+            ResolvedCatalogBaseTable<?> view,
+            ObjectIdentifier viewIdentifier,
+            boolean isTemporary) {
+        final String printIndent = "  ";
+        CatalogBaseTable.TableKind kind = view.getTableKind();
+        if (kind != CatalogBaseTable.TableKind.VIEW) {
+            throw new TableException(
+                    String.format(
+                            "SHOW CREATE VIEW does not support showing CREATE TABLE statement with identifier %s.",
```
Comment (Airblader) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java:1481]: I think we could make this more helpful for the user:

```suggestion
                            "SHOW CREATE VIEW is only supported for views, but '%s' is a table. Please use SHOW CREATE TABLE instead.",
```

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/catalog/CatalogViewITCase.scala (hunk 2 above)
Comment (RocMarshal) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/catalog/CatalogViewITCase.scala:38]: Good catch. Thank you. I'll remove it.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java (hunk 4 above)
Comment (RocMarshal) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java:1185]: OK. I'll update it.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/catalog/CatalogViewITCase.scala (hunk 7 above)
Comment (RocMarshal) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/catalog/CatalogViewITCase.scala]: Thanks for the comment. I updated it.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/catalog/CatalogViewITCase.scala (hunk 8 above)
Comment (RocMarshal) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/catalog/CatalogViewITCase.scala]: Good idea. 
Done.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java (hunk 5 above)
Comment (RocMarshal) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java:1514]: Good idea. 
Done.

**Comment by RocMarshal** [comment]
@Airblader Thank you for your comments. I made some changes based on your suggestions. Please let me know what's your opinion.

File: docs/content/docs/dev/table/sql/show.md [hunk 10]
```diff
@@ -26,7 +26,7 @@ under the License.
...
+SHOW statements are used to list all catalogs, or list all databases in the current catalog, or list all tables/views in the current catalog and the current database, or show current catalog and database, or show create statement for specified table or specified view, or list all functions including system functions and user-defined functions in the current catalog and current database, or list only user-defined functions in the current catalog and current database, or list enabled module names, or list all loaded modules with enabled status in the current session, or list the columns of the table or the view with the given name and the optional like clause.
```
Comment (Airblader) [inline docs/content/docs/dev/table/sql/show.md:29]: This isn't entirely related to your PR, but the sentence here is very lengthy and pretty much unreadable at this point. Can we maybe rephrase this a bit?

> `SHOW` statements are used to list objects within their corresponding parent, such as catalogs, databases, tables and views, columns, functions, and modules. See the individual commands for more details and additional options.
> `SHOW CREATE` statements are used to print a DDL statement with which a given object can be created.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java [hunk 11]
```diff
@@ -1362,35 +1392,31 @@ private TableResult buildShowResult(String columnName, String[] objects) {
...
-        sb.append(
-                schema.getColumns().stream()
-                        .map(column -> String.format("%s%s", printIndent, getColumnString(column)))
-                        .collect(Collectors.joining(",\n")));
-        // append watermark spec
-        if (!schema.getWatermarkSpecs().isEmpty()) {
-            sb.append(",\n");
-            sb.append(
-                    schema.getWatermarkSpecs().stream()
+    private String buildCreateFormattedPrefix(
```
Comment (Airblader) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java:1395]: Since this class is already very big, and this isn't really related to the table environment anymore, maybe it makes sense to move this formatting logic into some (`@Internal`, of course) utility class and declutter this class by doing so?

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java [hunk 12]
```diff
@@ -1403,44 +1429,120 @@ private String buildShowCreateTableRow(
...
+        if (Objects.isNull(table.getOptions()) || table.getOptions().isEmpty()) {
+            return Optional.empty();
+        }
+        return Optional.of(
+                table.getOptions().entrySet().stream()
+                        .map(
+                                entry ->
+                                        String.format(
+                                                "%s'%s' = '%s'",
+                                                printIndent, entry.getKey(), entry.getValue()))
```
Comment (Airblader) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java:1464]: The key and value need to be single-quote escaped as well.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java [hunk 13]
```diff
@@ -1403,44 +1429,120 @@ private String buildShowCreateTableRow(
...
 
+    /** Show create view statement only for views. */
+    private String buildShowCreateViewRow(
+            ResolvedCatalogBaseTable<?> view,
+            ObjectIdentifier viewIdentifier,
+            boolean isTemporary) {
+        if (view.getTableKind() != CatalogBaseTable.TableKind.VIEW) {
+            throw new TableException(
+                    String.format(
+                            "SHOW CREATE VIEW is only supported for views, but '%s' is a table. Please use SHOW CREATE TABLE instead.",
```
Comment (Airblader) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java:1521]: nit: I think this was my mistake, sorry, but removing the `''` around the `%s` here would probably be better since the identifier will be fully escaped anyway.

**Review by Airblader — COMMENTED** [review]
Thanks, this is looking much better already! I think we could maybe make a few more changes. What do you think?

**Comment by RocMarshal** [comment]
@Airblader Thank you so much for your suggestions. I updated the pr. PTAL.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/ShowCreateUtil.java [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/** SHOW CREATE statement Util. */
+@Internal
+enum ShowCreateUtil {
```
Comment (Airblader) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/ShowCreateUtil.java:41]: There's no need to make this an enum, it just looks confusing. Please just turn this into a class and add a private constructor.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/ShowCreateUtil.java [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+
+    static String extractFormattedColumns(ResolvedCatalogBaseTable<?> table, String printIndent) {
+        return table.getResolvedSchema().getColumns().stream()
+                .map(column -> String.format("%s%s", printIndent, getColumnString(column)))
+                .collect(Collectors.joining(",\n"));
+    }
+
+    static Optional<String> extractFormattedWatermarkSpecs(
+            ResolvedCatalogBaseTable<?> table, String printIndent) {
+        if (!table.getResolvedSchema().getWatermarkSpecs().isEmpty()) {
```
Comment (Airblader) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/ShowCreateUtil.java:93]: nit: we can invert this condition to exit early and reduce the indentation level:

```
if (table.getResolvedSchema().getWatermarkSpecs().isEmpty()) {
    return Optional.empty();
}

// …
```

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/ShowCreateUtil.java [hunk 16]
```diff
@@ -0,0 +1,221 @@
...
+        if (Objects.isNull(table.getOptions()) || table.getOptions().isEmpty()) {
+            return Optional.empty();
+        }
+        return Optional.of(
+                table.getOptions().entrySet().stream()
+                        .map(
+                                entry ->
+                                        String.format(
+                                                "%s\'%s\' = \'%s\'",
+                                                printIndent, entry.getKey(), entry.getValue()))
```
Comment (Airblader) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/ShowCreateUtil.java:140]: I don't understand what's going on here. Shouldn't this be this?

```suggestion
                                                "%s'%s' = '%s'",
                                                printIndent, EncodingUtils.escapeSingleQuotes(entry.getKey()), EncodingUtils.escapeSingleQuotes(entry.getValue())))
```

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/ShowCreateUtil.java [hunk 17]
```diff
@@ -0,0 +1,221 @@
...
+            throw new TableException(
+                    String.format(
+                            "SHOW CREATE TABLE is only supported for tables, but %s is a view. Please use SHOW CREATE VIEW instead.",
+                            tableIdentifier.asSerializableString()));
+        }
+        final String printIndent = "  ";
+        StringBuilder sb =
+                new StringBuilder()
+                        .append(buildCreateFormattedPrefix("TABLE", isTemporary, tableIdentifier));
+        // append columns
```
Comment (Airblader) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/ShowCreateUtil.java:165]: nit: all of these comments are redundant and noisy now because the method names already make it clear what's happening

**Review by Airblader — CHANGES_REQUESTED** [review]
Thanks @RocMarshal! This looks largely good to me now, minus a couple more comments of which most are nits. We'll need a committer to look at this now.

**Comment by RocMarshal** [comment]
Thanks @Airblader for the review. 
Hi, @wuchong , @xintongsong @twalthr . Could you help me to merge it if there are nothing inappropriate ? Thank you.

**Comment by RocMarshal** [comment]
@leonardBang Could you help me to merge it if there are nothing inappropriate ? Thank you.

**Comment by RocMarshal** [comment]
@wuchong I really appreciate it. I make some change based on your suggestions.
